packages/validators/src/schemas/v2023_09_28/experience.ts

```diff
@@ -0,0 +1,185 @@
+import { z } from 'zod';
+import { SchemaVersions } from '../schemaVersions';
+
+const uuidKeySchema = z.string().regex(/^[a-zA-Z0-9-_]{1,21}$/);
+const propertyKeySchema = z.string().regex(/^[a-zA-Z0-9-_]{1,64}$/);
+
+const DataSourceSchema = z.record(
+  uuidKeySchema,
+  z.object({
+    sys: z.object({
+      type: z.literal('Link'),
+      id: z.string(),
+      linkType: z.literal('Entry').or(z.literal('Asset')),
+    }),
+  }),
+);
+
+const PrimitiveValueSchema = z.union([
+  z.string(),
+  z.boolean(),
+  z.number(),
+  z.record(z.any(), z.any()),
+  z.undefined(),
+]);
+
+export const ComponentDefinitionPropertyTypeSchema = z.enum([
+  'Text',
+  'RichText',
+  'Number',
+  'Date',
+  'Boolean',
+  'Location',
+  'Media',
+  'Object',
+]);
+
+const ValuesByBreakpointSchema = z.record(z.lazy(() => PrimitiveValueSchema));
+
+const ComponentPropertyValueSchema = z.union([
+  z.object({
+    type: z.literal('DesignValue'),
+    valuesByBreakpoint: ValuesByBreakpointSchema,
+  }),
+  z.object({
+    type: z.literal('BoundValue'),
+    path: z.string(),
+  }),
+  z.object({
+    type: z.literal('UnboundValue'),
+    key: z.string(),
+  }),
+  z.object({
+    type: z.literal('ComponentValue'),
+    key: z.string(),
+  }),
+]);
+
+export type ComponentPropertyValue = z.infer<typeof ComponentPropertyValueSchema>;
+
+const BreakpointSchema = z.object({
+  id: propertyKeySchema,
+  query: z.string().regex(/^\*$|^<[0-9*]+px$/),
+  previewSize: z.string(),
+  displayName: z.string(),
+});
+
+const UnboundValuesSchema = z.record(
+  uuidKeySchema,
+  z.object({
+    value: PrimitiveValueSchema,
+  }),
+);
+
+// Use helper schema to define a recursive schema with its type correctly below
+const BaseComponentTreeNodeSchema = z.object({
+  definitionId: z.string(),
+  variables: z.record(propertyKeySchema, ComponentPropertyValueSchema),
+});
+export type ComponentTreeNode = z.infer<typeof BaseComponentTreeNodeSchema> & {
+  children: ComponentTreeNode[];
+};
+const ComponentTreeNodeSchema: z.ZodType<ComponentTreeNode> = BaseComponentTreeNodeSchema.extend({
+  children: z.lazy(() => ComponentTreeNodeSchema.array()),
+});
+
+const ComponentSettingsSchema = z.object({
+  variableDefinitions: z.record(
+    z.string().regex(/^[a-zA-Z0-9-_]{1,86}$/), // Here the key is <variableName>_<nanoid_id> so we need to allow for a longer length
```

Comment:
Are we already imposing any limits in the SDK, on custom components' variable definition?

If not, we should probably look into enforcing it at registration time already, to not end up violating anything indirectly as a consequence.

I think 16 or 32 would be a reasonable variable key length to start with.

Reply:
There is no real need for 64 char, we can go with 32 and we can relax this later if we see a need.
We don't have yet any enforcement on registration time, I'll create a small task to add that.

Reply:
Thank you!
This makes me even think that in the long run, we might want to reconsider the `<variableName>_<nanoid_id>` convention, there's a recursion problem if we think pattern and design system components should be interchangeable in principle :thinking: